### ISSUE TITLE
LDAP user import error

### ISSUE THREAD

**Issue by b3owulf101** [janeczku/calibre-web#2713]
**Describe the bug/problem**
When attempting to import LDAP users I get an error message of "Error: argument of type 'int' is not iterable"

**To Reproduce**
1. Set up LDAP configuration
2. Click on "Import LDAP Users" on admin screen

**Logfile**
[2023-03-06 05:29:01,158] DEBUG {cps.updater:329} Stable version: {'version': '0.6.19'}
[2023-03-06 05:29:02,420] ERROR {cps.admin:1577} argument of type 'int' is not iterable

**Environment (please complete the following information):**
 - OS: Ubuntu 20.04
 - Python version: python 3.10.6
 - Calibre-Web version: 0.6.19
 - Docker container: LinuxServer
 - Browser: Chrome and Firefox

**Additional context**
I am using Authentik's LDAP outpost. I have successfully used it with other applications.

**Comment by OzzieIsaacs** [comment]
I'm very bad with this LDAP stuff. I hope you understand that I can't install Authentik-Server to reproduce your issue, without any context. So which settings did you use, and for debugging it would be helpful for me to have a device tree, some something like (without passwords of course):
```
dn: uid=Dot.to,ou=people, dc=calibreweb,dc=com
objectClass: person
objectClass: organizationalPerson
objectClass: inetOrgPerson
uid: Dot.to
mail: dot@to.to
gn: Spanish
sn: Character
userPassword: password
#Generic groups
dn: ou=groups,dc=calibreweb,dc=com
objectclass:organizationalunit
ou: groups
# create the cps entry
dn: cn=cps,ou=groups,dc=calibreweb,dc=com
objectclass: groupofnames
cn: cps
member: uid=Mümmy 7,ou=People,dc=calibreweb,dc=com
member: uid=执一,ou=People,dc=calibreweb,dc=com
member: uid=Dot.to,ou=People,dc=calibreweb,dc=com
```

**Comment by b3owulf101** [comment]
I completely understand. I can share screenshots of my LDAP configuration.

![Screenshot_20230306-113609_1.png](https://user-images.githubusercontent.com/7886178/223174688-cf44ba72-8b0e-4c15-b473-5c40bca60bd7.png)
![Screenshot_20230306-113631_1.png](https://user-images.githubusercontent.com/7886178/223174734-62700815-6c40-448f-9014-d6bbc58c2ce7.png)

DC=Website was of course changed for privacy.
I am basing my config off of what I've seen other users post in the authentik discord and on this https://github.com/goauthentik/authentik/pull/4139

It doesn't matter what changes I make to the config (and I have tried putting in obviously wrong stuff like "not_available" into the fields 1 by 1) I always end up with the error message "Error: argument of type 'int' is not iterable". Except when I put in an incorrect server address, then I get an error message about being unable to reach the server.
![Screenshot_20230306-115246_1.png](https://user-images.githubusercontent.com/7886178/223177207-6eb72f10-4951-429a-89ff-5d51238a14b6.png)

The only log that gets generated when I click on import users is as listed above:
ERROR {cps.admin:1577} argument of type 'int' is not iterable

Which reads to me like it may be an error occurring in line 1577 of the admin.py script but I don't understand what exactly is going on in that bit of the script.

**Comment by OzzieIsaacs** [comment]
Sorry man, no device tree, no help

**Comment by b3owulf101** [comment]
Apologies. I forgot to include that. This is the result of running ldapsearch for a test account

```
# extended LDIF
#
# LDAPv3
# base <dc=ldap,dc=website,dc=info> with scope subtree
# filter: cn=TestACCT
# requesting: ALL
#

# TestAcct, users, ldap.website.info
dn: cn=TestAcct,ou=users,dc=ldap,dc=website,dc=info
uid: fbac98fb6b795704c63911234581e0e090fea933a49625c09cf988e031de8aa3
name: test account
objectClass: user
objectClass: organizationalPerson
objectClass: inetOrgPerson
objectClass: goauthentik.io/ldap/user
uidNumber: 2010
memberOf: cn=authentik-users,ou=groups,dc=ldap,dc=website,dc=info
memberOf: cn=calibreweb,ou=groups,dc=ldap,dc=website,dc=info
goauthentik.io/ldap/active: true
goauthentik.io/ldap/superuser: false
cn: TestAcct
sAMAccountName: TestAcct
displayName: test account
mail:
gidNumber: 2010

# TestAcct, virtual-groups, ldap.website.info
dn: cn=TestAcct,ou=virtual-groups,dc=ldap,dc=website,dc=info
cn: TestAcct
uid: fbac98fb6b795704c63911234581e0e090fea933a49625c09cf988e031de8aa3
sAMAccountName: TestAcct
gidNumber: 2010
objectClass: group
objectClass: groupOfUniqueNames
objectClass: groupOfNames
objectClass: goauthentik.io/ldap/group
objectClass: goauthentik.io/ldap/virtual-group
member: cn=TestAcct,ou=users,dc=ldap,dc=website,dc=info
goauthentik.io/ldap/superuser: false

# search result
search: 2
result: 0 Success

# numResponses: 3
# numEntries: 2
```

**Comment by OzzieIsaacs** [comment]
I can't reproduce the error. But your config doesn't match the tree you send me:
The name of the group you try to import is: "TestAcct". Wasn't matching:
![grafik](https://user-images.githubusercontent.com/6438895/227765405-2c01a52c-b082-4e62-aeb2-a6f8eda12214.png)
Of course this can be a mistake while generating the example.
I'm generating using a python ldap server (so I can influence everything and don't have to deal with different servers).
I extended the device tree you send me to get a fully example to this one:
```
dn: dc=info
dc: info
objectClass: dcObject

dn: dc=website,dc=info
dc: website
objectClass: dcObject
objectClass: organization

dn: dc=ldap,dc=website,dc=info
dc: ldap
objectClass: dcObject
objectClass: SubOrganization

dn: ou=users,dc=ldap,dc=website,dc=info
objectClass: organizationalUnit
ou: users

#Generic groups
dn: ou=virtual-groups,dc=ldap,dc=website,dc=info
objectclass:organizationalunit
ou: virtual-groups

dn: cn=root,dc=ldap,dc=website,dc=info
cn: root
gn: root
mail: admin@calibreweb.com
objectclass: top
objectclass: person
objectClass: inetOrgPerson
sn: admin
userPassword: secret

# TestAcct, users, ldap.website.info
dn: cn=TestAcct,ou=users,dc=ldap,dc=website,dc=info
uid: fbac98fb6b795704c63911234581e0e090fea933a49625c09cf988e031de8aa3
name: test account
objectClass: user
objectClass: organizationalPerson
objectClass: inetOrgPerson
objectClass: goauthentik.io/ldap/user
uidNumber: 2010
memberOf: cn=authentik-users,ou=groups,dc=ldap,dc=website,dc=info
memberOf: cn=calibreweb,ou=groups,dc=ldap,dc=website,dc=info
goauthentik.io/ldap/active: true
goauthentik.io/ldap/superuser: false
cn: TestAcct
sAMAccountName: TestAcct
displayName: test account
mail:
gidNumber: 2010

# TestAcct, virtual-groups, ldap.website.info
dn: cn=TestAcct,ou=virtual-groups,dc=ldap,dc=website,dc=info
cn: TestAcct
uid: fbac98fb6b795704c63911234581e0e090fea933a49625c09cf988e031de8aa3
sAMAccountName: TestAcct
gidNumber: 2010
objectClass: group
objectClass: groupOfUniqueNames
objectClass: groupOfNames
objectClass: goauthentik.io/ldap/group
objectClass: goauthentik.io/ldap/virtual-group
member: cn=TestAcct,ou=users,dc=ldap,dc=website,dc=info
goauthentik.io/ldap/superuser: false
userPassword: test
```

**Comment by OzzieIsaacs** [comment]
An option to get more information is the following. Please locate the `__init__.py` file from flask_simple_ldap inside your docker container. 
Go to line 95 and replace:
```
            conn = ldap.initialize('{0}://{1}:{2}'.format(
                current_app.config['LDAP_SCHEMA'],
                current_app.config['LDAP_HOST'],
                current_app.config['LDAP_PORT']))
```
```
with:
            conn = ldap.initialize('{0}://{1}:{2}'.format(
                current_app.config['LDAP_SCHEMA'],
                current_app.config['LDAP_HOST'],
                current_app.config['LDAP_PORT']), trace_level=2)
```
(last line the trace_level=2). Afterwards restart calibre-web and try to import the users again, there should be lot more output like:
```
*** <ldap.ldapobject.SimpleLDAPObject object at 0x7f42f2b135b0> ldap://127.0.0.1:3268 - SimpleLDAPObject.result4
((2, 1, -1, 0, 0, 0), {})
=> result:
(101,
 [('cn=TestAcct,ou=virtual-groups,dc=ldap,dc=website,dc=info',
   {'cn': [b'TestAcct'],
    'gidNumber': [b'2010'],
    'goauthentik.io/ldap/superuser': [b'false'],
    'member': [b'cn=TestAcct,ou=users,dc=ldap,dc=website,dc=info'],
    'objectClass': [b'goauthentik.io/ldap/group',
                    b'goauthentik.io/ldap/virtual-group',
                    b'group',
                    b'groupOfNames',
                    b'groupOfUniqueNames'],
    'sAMAccountName': [b'TestAcct'],
    'uid': [b'fbac98fb6b795704c63911234581e0e090fea933a49625c09cf988e031de8aa3'],
    'userPassword': [b'test']}),
  ('cn=TestAcct,ou=users,dc=ldap,dc=website,dc=info',
   {'cn': [b'TestAcct'],
    'displayName': [b'test account'],
    'gidNumber': [b'2010'],
    'goauthentik.io/ldap/active': [b'true'],
    'goauthentik.io/ldap/superuser': [b'false'],
    'mail': [b''],
    'memberOf': [b'cn=authentik-users,ou=groups,dc=ldap,dc=website,dc=info',
                 b'cn=calibreweb,ou=groups,dc=ldap,dc=website,dc=info'],
    'name': [b'test account'],
    'objectClass': [b'goauthentik.io/ldap/user',
                    b'inetOrgPerson',
                    b'organizationalPerson',
                    b'user'],
    'sAMAccountName': [b'TestAcct'],
    'uid': [b'fbac98fb6b795704c63911234581e0e090fea933a49625c09cf988e031de8aa3'],
    'uidNumber': [b'2010']})],
 2,
 [])
*** <ldap.ldapobject.SimpleLDAPObject object at 0x7f42f2b135b0> ldap://127.0.0.1:3268 - SimpleLDAPObject.unbind_ext
((None, None), {})
=> result:
None
[2023-03-26 11:07:37,015]  WARN {cps.admin:1548} LDAP User  {'objectClass': [b'goauthentik.io/ldap/group', b'goauthentik.io/ldap/virtual-group', b'group', b'groupOfNames', b'groupOfUniqueNames'], 'cn': [b'TestAcct'], 'gidNumber': [b'2010'], 'goauthentik.io/ldap/superuser': [b'false'], 'member': [b'cn=TestAcct,ou=users,dc=ldap,dc=website,dc=info'], 'sAMAccountName': [b'TestAcct'], 'uid': [b'fbac98fb6b795704c63911234581e0e090fea933a49625c09cf988e031de8aa3'], 'userPassword': [b'test']} Already in Database
```

**Comment by OzzieIsaacs** [comment]
This issue had the same error message: #1733

**Comment by b3owulf101** [comment]
That tree I gave you was for an account, not a group. the group I'm using is calibreweb. The results of ldapsearch are below:
```
# extended LDIF
#
# LDAPv3
# base <dc=ldap,dc=website,dc=info> with scope subtree
# filter: cn=calibreweb
# requesting: ALL
#

# calibreweb, groups, ldap.website.info
dn: cn=calibreweb,ou=groups,dc=ldap,dc=website,dc=info
sAMAccountName: calibreweb
gidNumber: 26931
objectClass: group
objectClass: groupOfUniqueNames
objectClass: groupOfNames
objectClass: goauthentik.io/ldap/group
member: cn=collin,ou=users,dc=ldap,dc=website,dc=info
member: cn=TestAcct,ou=users,dc=ldap,dc=website,dc=info
goauthentik.io/ldap/superuser: false
cn: calibreweb
uid: 11407f27-78bb-4d77-b197-96e2731833d8

# search result
search: 2
result: 0 Success

# numResponses: 2
# numEntries: 1
```
I added the "trace_level=2" to where you specified. In the linuxserver container it's located at `/usr/local/lib/python3.10/dist-packages/flask_simpleldap` and I restarted calibre-web from the admin page. I tried importing users again and got the same error message. The logs show the same error message as before `[2023-03-26 06:01:22,125] ERROR {cps.admin:1577} argument of type 'int' is not iterable`. I'm not seeing anything extra getting displayed. I am checking the log messages from the calibre-web admin page and log level is set to debug. Is there a different place I should be looking for logs or a specific logfile?

**Comment by OzzieIsaacs** [comment]
My fault, the output goes always to stdout (normal console output, let me check)

**Comment by OzzieIsaacs** [comment]
Is it easily possible to redirect calibre-web's console output in the container to a file?
(The required changes are otherwise a bit bigger)

**Comment by OzzieIsaacs** [comment]
Please try to edit the same file as before (lines 90-95):

```
        f_o = open("calibre-web.log", "a")
        try:
            conn = ldap.initialize('{0}://{1}:{2}'.format(
                current_app.config['LDAP_SCHEMA'],
                current_app.config['LDAP_HOST'],
                current_app.config['LDAP_PORT']), trace_level=log_level, trace_file=f_o)
```
Restart calibre-web, afterwards the additional outputs should get logged to calibre-web.log file (you can name the log file whatever you want). For me it's working, but I can't guarantee it for the container

**Comment by b3owulf101** [comment]
Thank you for that. I modified it as stated and got:

```
[2023-03-26 07:17:13,507] ERROR {cps:1457} Exception on /import_ldap_users [POST]
Traceback (most recent call last):
  File "/usr/local/lib/python3.10/dist-packages/flask/app.py", line 2073, in wsgi_app
    response = self.full_dispatch_request()
  File "/usr/local/lib/python3.10/dist-packages/flask/app.py", line 1518, in full_dispatch_request
    rv = self.handle_user_exception(e)
  File "/usr/local/lib/python3.10/dist-packages/flask/app.py", line 1516, in full_dispatch_request
    rv = self.dispatch_request()
  File "/usr/local/lib/python3.10/dist-packages/flask/app.py", line 1502, in dispatch_request
    return self.ensure_sync(self.view_functions[rule.endpoint])(**req.view_args)
  File "/usr/local/lib/python3.10/dist-packages/flask_login/utils.py", line 303, in decorated_view
    return current_app.ensure_sync(func)(*args, **kwargs)
  File "/app/calibre-web/cps/admin.py", line 95, in inner
    return f(*args, **kwargs)
  File "/app/calibre-web/cps/admin.py", line 1575, in import_ldap_users
    new_users = services.ldap.get_group_members(config.config_ldap_group_name)
  File "/app/calibre-web/cps/services/simpleldap.py", line 96, in get_group_members
    return _ldap.get_group_members(group)
  File "/usr/local/lib/python3.10/dist-packages/flask_simpleldap/__init__.py", line 256, in get_group_members
    conn = self.bind
  File "/usr/local/lib/python3.10/dist-packages/flask_simpleldap/__init__.py", line 115, in bind
    conn = self.initialize
  File "/usr/local/lib/python3.10/dist-packages/flask_simpleldap/__init__.py", line 95, in initialize
    current_app.config['LDAP_PORT']), trace_level=log_level, trace_file=f_o)
NameError: name 'log_level' is not defined
```
I then changed log_level to 2 and got the same error message as before in calibre-web.log. I then changed the name of the log file to calibre-web-debug.log, restarted calibre, ran import users, then opened the file named calibre-web-debug.log and it was empty. Nothing at all.

**Comment by OzzieIsaacs** [comment]
> I then changed log_level to 2 and got the same error message as before in calibre-web.log.

Sorry, again. Absolutly right, I mixed it up during some of my experiments.

I start believing this has to do something with docker. Would it generate big trouble for you to install calibre-web directly in a virtual environment, do the config and try if it's easier to get the debug message?

**Comment by OzzieIsaacs** [comment]
Okay stop

**Comment by OzzieIsaacs** [comment]
Another idea, if the file is empty then the initial call (the lines we are changed lead to the crash.

**Comment by OzzieIsaacs** [comment]
Another idea, if the file is empty then the initial call (the lines we are changed lead to the crash.
The problem is:
![grafik](https://user-images.githubusercontent.com/6438895/227773633-246597f8-1299-4ccf-ba34-b91a038f890d.png)
**You are supposed to enter "0.0.0.0" without the ldap**
Sorry this isn't catched

**Comment by b3owulf101** [comment]
I got it working. Details in case anybody else happens upon this:

I removed the LDAP:// and changed the hostname to `ak-outpost-ldap` which is the default name of authentik's LDAP outpost container and made sure both calibre-web and the outpost container were on the same docker network. I also changed the port to 3389 instead of 389. Authentik by default maps the container port 389 to host port 3389. I thought since this was part of the same docker network port 389 would be the port to use, I was wrong. I thought I had previously tried both 389 and 3389 but I guess I was mistaken.

Thanks for the help OzzieIsaacs. I apologize for this convoluted mess when there was a simple config error lol.
